src/net_processing.cpp

```diff
@@ -1433,7 +1435,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
     // Use deterministic randomness to send to the same nodes for 24 hours
     // at a time so the m_addr_knowns of the chosen nodes prevent repeats
     uint64_t hashAddr = addr.GetHash();
-    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write((GetTime() + hashAddr) / (24 * 60 * 60));
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = static_cast<uint64_t>((current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL);
```

Comment:
We iterated over this line a couple times with @MarcoFalke, so I'm out of ideas here.
Open to suggestions!

Reply:
Annoyingly, I think this is UB, as hashAddr may exceed 2^63, and casting a larger number to a signed 64-bit integer would be outside its range (and overflowing signed integers is undefined behavior).

Suggestion: `uint64_t time_addr = (current_time + std::chrono::seconds{hashAddr % ROTATE_ADDR_RELAY_DEST_INTERVAL.count()}) / ROTATE_ADDR_RELAY_DEST_INTERVAL;`, and make `ROTATE_ADDR_RELAY_DEST_INTERVAL` to be of type `std::chrono::seconds`.

Reply:
This isn't resolved actually; it's still casting `hashAddr` from `uint64_t` to `int64_t`, which is implementation-defined behavior (I was wrong claiming earlier that it was UB).

Reply:
We're constructing a `std::chrono::seconds` from an `int64_t`, but according to https://en.cppreference.com/w/cpp/chrono/duration, `std::chrono::seconds` is only guaranteed to have 35 bits -- is this problematic?

Reply:
Uh, right. I think that could be a problem. This will need something like

```c++
uint64_t time_addr = (uint64_t{current_time / std::chrono::seconds{1}} + hashAddr) / (ROTATE_ADDR_RELAY_DEST_INTERVAL / std::chrono::seconds{1});
```

(I prefer using `x / std::chrono::seconds{1}` over `x.count()` to get the number of seconds of `x`, as the former expression remains meaningful if the type of `x` changes to something other than `std::chrono::seconds`).

Reply:
My understanding was that `std::chrono::seconds{int64_t{}}` wouldn't compile on platforms where seconds isn't backed by at least 64 bits. I don't expect any such platform to exist.

@ sipa there is also a `count_seconds()` helper.

Reply:
Github is weird, I found this un-resolved issue totally accidentally...

Reply:
Okay the latest version should be safe?